Add support for --extra-attributes to Gazetteer output

Nakaner opened the pull request:
`--extra-attributes` is now supported by Gazetteer output. This pull requests writes all metadata fields to as to the `extratags` column.

If an OSM object lacks a metadata field (empty string or 0 as value) which is/will be common with usual OSM planet dumps and extracts, the field will not be written to the output.

Users can control whether they want all metadata fields or only specific ones. If the Nominatim style file (the new one, see #1277) contains a catch-all rule at the end, that catch-all rule will apply on all available metadata fields and they will be written to the `places` table. If one or many of the metadata fields are mentioned in an `extra` rule and no catch-all rule is present, only that field will be written.

**Example 1: write all tags and all metadata fields to the `extratags` column**

Append this to your style file:

```json
{
    "keys" : [""],
    "values" : {
        "" : "extra"
    }
}
```

**Example 2: write all metadata fields to the `extratags` column**

Append this to your style file:

```json
{
    "keys" : ["osm_version", "osm_timestamp", "osm_user", "osm_uid", "osm_changeset"],
    "values" : {
        "" : "extra"
    }
}
```

**Example 3: write timestamps to the `extratags` column**

Append this to your style file:

```json
{
    "keys" : ["osm_timestamp"],
    "values" : {
        "" : "extra"
    }
}
```

lonvia commented on gazetteer-style.cpp:
```diff
@@ -502,6 +547,13 @@ bool gazetteer_style_t::copy_out_maintag(pmaintag_t const &tag,
         }
+        for (auto const &entry : m_metadata) {
+            buffer += "\"";
+            escape_array_record(entry.first.c_str(), buffer);
```
You can save the escaping here. All entry.first are well-known strings.

lonvia commented on gazetteer-style.hpp:
```diff
@@ -12,6 +12,7 @@ class gazetteer_style_t
 {
     using flag_t = uint16_t;
     using ptag_t = std::pair<char const *, char const *>;
+    using ptag_str_t = std::pair<const std::string, const std::string>;
```
'char const *' will do for the first parameter. It is only used with constant strings.

lonvia commented on gazetteer-style.hpp:
```diff
@@ -64,11 +65,17 @@ class gazetteer_style_t
+    void add_metadata_field(const std::string&& field, const std::string&& value);
+
+    template <typename T>
+    void add_metadata_field_num(const std::string&& field, const T value);
```
Please use the const placement style as in the rest of the file.

lonvia commented on gazetteer-style.cpp:
```diff
@@ -384,6 +390,45 @@ void gazetteer_style_t::process_tags(osmium::OSMObject const &o)
+    }
+}
+
+void gazetteer_style_t::add_metadata_field(const std::string&& field, const std::string&& value) {
```
Please run clang-format over your change (see CONTRIBUTING.md).

lonvia commented on gazetteer-style.cpp:
```diff
@@ -384,6 +390,45 @@ void gazetteer_style_t::process_tags(osmium::OSMObject const &o)
+
+template <typename T>
+void gazetteer_style_t::add_metadata_field_num(const std::string&& field, const T value) {
+    // This method is not linked to from outside this class. Therefore, it can stay in the source file.
```
I'm not sure, I'm comfortable with that. Better move it to the header.

lonvia commented on gazetteer-style.cpp:
```diff
@@ -163,6 +163,25 @@ gazetteer_style_t::flag_t gazetteer_style_t::parse_flags(std::string const &str)
 }
 
+bool gazetteer_style_t::add_metadata_style_entry(std::string const &key,
+                                                 std::string const &value)
```
'value' is unused here. You can just remove the second parameter. (I wonder why travis does not complain, we need to change that.)

Nakaner commented:
I added a check to ensure that only the "extra" flag is set.

I rebased on current master and adapted my code to your recent changes implemented in #914.
